Add generate-themes-doc.fish

Pyppe opened the pull request:
Wanted to browse thru all the available themes, but could not find a centralized place for this. This PR contains a `bin/generate-themes-doc.fish` which generates `docs/Themes.md` that can be used for previewing the themes.

See https://github.com/Pyppe/oh-my-fish/blob/master/docs/Themes.md

bpinto commented:
I love this!!! Is there a way to automate it though?

derekstavis commented:
Thanks!! I like having this docs!! By the way, the script could be improved by generating a table of contents on the top of the docs. What do you think about?

Pyppe commented:
Thanks for the feedback, guys! I improved the script, and added TOC.

Regarding automation; I don't have a silver bullet. Of course a custom script could be added that overrides the generated `docs/Themes.md` and pushes it to master if it was changed... And then this could be invoked e.g. for each Travis CI-build. But personally I wouldn't want this kind of automagic behavior for my project... :)

Hmm... after-thought: Travis could generate the `Themes.md` and send email if it has changed. (Although, I don't know how easy/possible this kind of automation is to implement in Travis).

Pyppe commented:
Regarding automation, I think this could be one solution https://github.com/Pyppe/oh-my-fish/commit/a898f84691dcf133c0783d33fb1ec0a55db6a3b7: It runs the `generate-themes-doc.fish` as a Travis-test, and the build fails if `Themes.md` has changed.

Travis seems to be having some issues (https://twitter.com/pyrylahti/status/645578947560873985); so I'm yet to verify that this actually works...

derekstavis commented:
Looks awesome with the TOC!! I was also wondering if we should include a "Back to top" link after every theme, as navigation gets hard on such big files.

derekstavis commented:
Regarding to the place where the script is stored, maybe we should add this category of scripts to a `tools` directory, @oh-my-fish/core?

bpinto commented:
I like using Travis for the automation! Very clever. Maybe tools is a good folder.

derekstavis commented:
So I propose the following:

    bin/generate-themes-docs.fish -> tools/generate-themes-docs.fish

As I believe that PR #52 will also produce some other scripts that fit into `tools` directory.

Pyppe commented:
* Changed the directory from `bin/` to `tools/`
* Updated `README.md` to mention `Themes.md`

IMO we could already merge this as is. I can add the Travis integration later once #52 gets merged (because it seems to do some of the same things I'd like to do).

bpinto commented on tools/generate-themes-doc.fish:
```diff
@@ -0,0 +1,55 @@
+
+set -l project_dir (status -f|xargs dirname|xargs dirname)
+set -l theme_doc "$project_dir/docs/Themes.md"
+set temp_theme_contents (mktemp /tmp/fish.Themes.Content.XXXXX.md)
```
Does mktemp work on Linux and OS X?

Pyppe replied:
IMO, yes.

bpinto commented on tools/generate-themes-doc.fish:
```diff
@@ -0,0 +1,55 @@
+#!/usr/bin/fish
```
Get fish binary using env?

bpinto commented on tools/generate-themes-doc.fish:
```diff
@@ -0,0 +1,55 @@
+set temp_theme_contents (mktemp /tmp/fish.Themes.Content.XXXXX.md)
+set temp_theme_toc (mktemp /tmp/fish.Themes.TOC.XXXXX.md)
+
+function write_theme_readme -a name raw_content readme
```
Can we follow the guidelines for private functions?

bpinto commented on tools/generate-themes-doc.fish:
```diff
@@ -0,0 +1,55 @@
+echo "Generating Themes documentation ..."
+echo "# Available themes" > $temp_theme_toc
+
+for t in (command find $project_dir/db/themes/ -type f|sort)
```
Theme instead of t?

bpinto commented on tools/generate-themes-doc.fish:
```diff
@@ -0,0 +1,55 @@
+    echo "Fetching readme for $name"
+    write_theme_readme $name $raw_content $readme
+  else
+    echo "FAILED: No readme for $name"
```
Output to stderr?

Pyppe commented:
Made changes. Some questions, though:
* What do you mean by "Get fish binary using env?"
* Renamed scripts functions with `__` prefix? Is this enough / OK?

bpinto commented:
The prefix is perfect! Regarding env, this is what I mean: http://stackoverflow.com/questions/2429511/why-do-people-write-usr-bin-env-python-on-the-first-line-of-a-python-script

Pyppe commented:
Roger. Changed the shebang-line to `#!/usr/bin/env fish`

bpinto commented:
Really nice!!! Eager to see the automation using Travis and perhaps a similar feature for plugins?

:+1: for the merge. Would you mind squashing your commits into one?

derekstavis commented:
I've tested here, no issues. :+1: for the merge. Also, rebasing into current master would be great to keep the history linear :)

Pyppe commented:
Rebased and squashed. :shipit:

derekstavis commented:
Super!! Thanks @Pyppe!!

bpinto commented:
@Pyppe we are back. Whenever you have a chance to work on the travis script, submit a PR! :dancers:

Pyppe commented:
I've got the functionality here: https://github.com/Pyppe/oh-my-fish/commit/de464e56b1b401b5e9bc83f604ffe16b4d991244

However, I think we'd need some changes from the #52 PR in order to run these fish-based scripts within Travis. See https://travis-ci.org/Pyppe/oh-my-fish/builds/82388150

bpinto commented:
It has been merged! :dancers:
